libtransact/src/scheduler/serial/core.rs

```diff
@@ -156,40 +175,147 @@ impl SchedulerCore {
+                    TransactionExecutionResult::Invalid(InvalidTransactionResult {
+                        transaction_id: receipt.transaction_id.clone(),
+                        error_message: format!(
+                            "Containing batch ({}) is invalid",
+                            current_batch_id,
+                        ),
+                        error_data: vec![],
+                    })
+                }
+                TransactionExecutionResult::Invalid(_) => {
+                    // When an invalid transaction is encountered, the scheduler should fail-fast
+                    // and invalidate the whole batch immediately; if an invalid result is in the
+                    // previously executed transaction results, this did not happen.
+                    panic!("Should not already have an invalid result");
                 }
-            };
-
-            let context_id = match self.previous_context {
-                Some(previous_context_id) => self
-                    .context_lifecycle
-                    .create_context(&[previous_context_id], &self.state_id),
-                None => self.context_lifecycle.create_context(&[], &self.state_id),
-            };
-
-            self.current_batch = Some(pair);
-            self.execution_tx
-                .send(ExecutionTask::new(transaction, context_id))?
+            })
+            .collect();
+
+        self.txn_results
+            .push(TransactionExecutionResult::Invalid(invalid_result));
+
+        // Invalidate all unexecuted transactions in the batch
+        self.txn_results.append(
+            &mut self
+                .txn_queue
+                .drain(..)
+                .map(|txn| {
+                    TransactionExecutionResult::Invalid(InvalidTransactionResult {
+                        transaction_id: txn.header_signature().into(),
+                        error_message: format!(
+                            "Containing batch ({}) is invalid",
+                            current_batch_id
+                        ),
+                        error_data: vec![],
+                    })
+                })
+                .collect(),
+        );
+
+        Ok(())
+    }
+
+    fn send_batch_result(&mut self) -> Result<(), CoreError> {
+        let batch = self.current_batch.take().ok_or_else(|| {
+            CoreError::InternalError(
+                "attempting to send batch result but no current batch is executing".into(),
+            )
+        })?;
+
+        let mut results = vec![];
+        std::mem::swap(&mut results, &mut self.txn_results);
+
+        let batch_result = BatchExecutionResult { batch, results };
+
+        let shared = self
+            .shared_lock
+            .lock()
+            .expect("scheduler shared lock is poisoned");
+        match shared.result_callback() {
+            Some(callback) => {
+                callback(Some(batch_result));
+            }
+            None => {
+                warn!(
```

Comment:
I think this would be slightly more elegant to make put this into a function, and then add it as a callback until another was set.

Reply:
This will be addressed in the error callback PR